core/src/dsnp/graph_page.rs

```diff
@@ -0,0 +1,215 @@
+		Self { pages: HashMap::<PageId, GraphPage>::new() }
+	}
+
+	/// Getter for Pages in Graph
+	pub fn pages(&self) -> &HashMap<PageId, GraphPage> {
+		&self.pages
+	}
+
+	/// Setter for Pages in Graph
+	pub fn set_pages(&mut self, pages: HashMap<PageId, GraphPage>) {
+		self.pages = pages;
+	}
+
+	/// Create a new Page in the Graph, with the given PageId.
+	///
+	/// Error on duplicate PageId.
+	/// If Some(Page) supplied, insert the given page.
+	/// Otherwise, create a new empty page.
+	pub fn create_page(
+		&mut self,
+		page_id: &PageId,
+		page: Option<GraphPage>,
+	) -> Result<&mut GraphPage, &str> {
+		if let Some(_existing_page) = self.pages.get(page_id) {
+			return Err("Attempt to create a new page for an existing page ID")
+		}
+
+		self.pages.insert(
+			*page_id,
+			match page {
+				Some(page) => page,
+				None => GraphPage::new(),
+			},
+		);
+		Ok(self
+			.pages
+			.get_mut(page_id)
+			.expect("Unable to retrieve graph page just inserted"))
+	}
+
+	/// Retrieve the page with the given PageId
+	pub fn get_page(&self, page_id: &PageId) -> Option<&GraphPage> {
+		self.pages.get(page_id)
+	}
+
+	/// Retrieve a mutable reference to the page with the given PageId
+	pub fn get_page_mut(&mut self, page_id: &PageId) -> Option<&mut GraphPage> {
+		self.pages.get_mut(page_id)
+	}
+
+	/// Return the PageId in which the given connection resides, if found.
+	pub fn find_connection(&self, dsnp_id: &DsnpUserId) -> Option<PageId> {
+		for (id, page) in self.pages.iter() {
+			if page.contains(dsnp_id) {
+				return Some(*id)
+			}
+		}
+
+		None
+	}
+
+	/// Add a connection to the specified page.
+	/// This is used internally by the Graph Update Manager or Import
+	/// If the specified page does not exist, a new page will be created
+	/// and the connection inserted into it.
+	pub fn add_connection_to_page<'a>(
+		&'a mut self,
+		page_id: &PageId,
+		connection_id: &DsnpUserId,
+	) -> Result<(), &str> {
+		if !self.pages.contains_key(page_id) {
+			self.pages.insert(*page_id, GraphPage::new());
+		}
+		let page = self.get_page_mut(page_id).expect("Unable to retrieve page");
+		page.add_connection(connection_id)
+	}
+
+	/// Remove a connection from the graph.
+	/// Returns Ok(Option<PageId>) containing the PageId of the page
+	/// the connection was removed from, or Ok(None) if the connection
+	/// was not found.
+	pub fn remove_connection(
+		&mut self,
+		connection_id: &DsnpUserId,
+	) -> Result<Option<PageId>, &str> {
+		if let Some(page_id) = self.find_connection(connection_id) {
+			return match self.get_page_mut(&page_id) {
+				Some(page) => match page.remove_connection(connection_id) {
+					Ok(()) => Ok(Some(page_id)),
+					Err(e) => Err(e),
+				},
+				None => Err("Unable to retrieve page"),
+			}
+		}
+
+		// Return Ok if no-op/connection not found
+		Ok(None)
+	}
+}
+
+/// Macro to get an iterator to all connections across all GraphPages
+/// within a Graph.
+#[macro_export]
+macro_rules! iter_graph_connections {
```

Comment:
I like defining macros since they are very cool. Are there any advantages with macros over having a method on graph that does the same thing?

Reply:
I wrote this initially as a function, but ran afoul of lifetime/ownership/sharing issues when trying to return a flattened iterator from a closure. The macro got around the issue.